src/lib/homestore/heap_chunk_selector.h

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include <homestore/chunk_selector.h>
+#include <homestore/vchunk.h>
+#include <homestore/homestore_decl.hpp>
+
+#include <queue>
+#include <vector>
+#include <mutex>
+#include <utility>
+#include <memory>
+
+#include <folly/concurrency/ConcurrentHashMap.h>
+
+namespace homeobject {
+
+using csharedChunk = homestore::cshared<homestore::Chunk>;  
+
+class HeapChunkSelector : public homestore::ChunkSelector {
+public:
+    HeapChunkSelector() = default;
+    HeapChunkSelector(const HeapChunkSelector&) = delete;
+    HeapChunkSelector(HeapChunkSelector&&) noexcept = delete;
+    HeapChunkSelector& operator=(const HeapChunkSelector&) = delete;
+    HeapChunkSelector& operator=(HeapChunkSelector&&) noexcept = delete;
+    ~HeapChunkSelector() = default;
+
+    using VChunk = homestore::VChunk;
+
+    class VChunkComparator {
+    public:
+        bool operator()(VChunk& lhs, VChunk& rhs) {
+            return lhs.available_blks() < rhs.available_blks();
+        }
+    };
+
+    using VChunkHeap = std::priority_queue<VChunk, std::vector<VChunk>, VChunkComparator>;
+
+    void add_chunk(csharedChunk&) override;
+    void foreach_chunks(std::function< void(csharedChunk&) >&& cb) override;
```

Comment:
This too; are we intending the closure to mutate the Chunk Pointer itself? Or just the Chunk....